src/main/java/hudson/plugins/sshslaves/SSHLauncher.java

```diff
@@ -592,7 +594,30 @@ private void startAgent(SlaveComputer computer, final TaskListener listener, Str
                             String workingDirectory) throws IOException {
         session = connection.openSession();
         expandChannelBufferSize(session,listener);
-        String cmd = "cd \"" + workingDirectory + "\" && " + java + " " + getJvmOptions() + " -jar " + AGENT_JAR +
+
+        PrintStream logger = listener.getLogger();
+
+         // Code to support alternate drives (such as D:\) for the remote root directory on Windows, when not using cygwin.
+         // You can disable this feature with startup option: -Dhudson.plugins.sshslaves.alternativewindowsdrive=false
+        boolean AlternativeWindowsDriveCheck = ! "false".equals(System.getProperty("hudson.plugins.sshslaves.alternativewindowsdrive"));
+        String agentOSTYPE="";
+        try {
+            ByteArrayOutputStream agentOsTypeStream = new ByteArrayOutputStream();
+            connection.exec("echo \"$OSTYPE\"", agentOsTypeStream);
+            agentOSTYPE = StringUtils.chomp( agentOsTypeStream.toString(Charset.defaultCharset().name()) );
+        } catch (InterruptedException ex) {
+            // exceptions here will only disable the AlternativeWindowsDriveCheck
+            logger.println(Messages.SSHLauncher_UnableToGetEnvironment(getTimestamp()));
+            AlternativeWindowsDriveCheck=false;
+        }
+        Pattern windowsDrivePattern = Pattern.compile("^[a-zA-Z]:\\\\");
+        Matcher windowsDriveMatcher = windowsDrivePattern.matcher(workingDirectory);
+        String cdArguments = "";
+        if ( AlternativeWindowsDriveCheck && ! "cygwin".equals(agentOSTYPE) && windowsDriveMatcher.find() ) {
+            cdArguments = "/d ";
+        }
+
+        String cmd = "cd " + cdArguments + "\"" + workingDirectory + "\" && " + java + " " + getJvmOptions() + " -jar " + AGENT_JAR +
```

Comment:
Now I know you didn't cause this, but it would be nice to ensure that arguments are properly quoted when constructing raw commands like this. However, since this script seems to be agnostic of bash or batch, I'm not exactly sure what would be the best way to ensure that happens here.

For comparison, the easiest way to ensure user input is quoted properly in a command is to rely on the underlying environment to do that for you. In that case, it'd be to set environment variables with the user input, then executing a shell script that refers to properly quoted forms of the environment variables. That way the shell handles the escaping for you.

For some code you can look into, see https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/util/WindowsUtil.java for the Windows side of things at least.

Reply:
I think that is a good idea to address in a separate PR